File: desktop/electron/pipeline.ts
import type { Db } from './db';
import { detectText } from './detect';
import type { ClipboardEvent } from './clipboard-poll';
import type { ContentType } from './ipc-types';

export interface PipelineDeps {
  db: Db;
  excludedApps: string[];
  historySize: number;
  getFocusedApp: () => string | null;
  onNewClip: (id: number, contentType: ContentType) => void;
}

export function makeHandler(deps: PipelineDeps): (e: ClipboardEvent) => void {
  return (ev) => {
    const focused = deps.getFocusedApp();
    if (focused && deps.excludedApps.some((a) => a.toLowerCase() === focused.toLowerCase())) {
      return;
    }
    if (ev.kind === 'text') {
      const ct = detectText(ev.content, focused);
      const preview = ev.content.slice(0, 280);
      try {
        const ins = deps.db.insertClip(
          ct,
          Buffer.from(ev.content, 'utf8'),
          ev.mime,
          preview,
          focused,
          Date.now()
        );
        if (ins.wasNew) {
          deps.onNewClip(ins.id, ct);
          deps.db.prune(deps.historySize);
        }
      } catch (e) {
        console.warn('[pipeline] text insert failed', e);
      }
      return;
    }
    if (ev.kind === 'image') {
      const preview = `Image ${ev.pngBytes.length} bytes`;
      try {
        const ins = deps.db.insertClip(
          'image',
          ev.pngBytes,
          'image/png',
          preview,
          focused,
          Date.now()
        );
        if (ins.wasNew) {
          // thumbnail: store the PNG itself (Electron renders it fine; resize
          // can come later via Sharp if we want true 200x200 thumbnails)
          deps.db.setThumbnail(ins.id, ev.pngBytes);
          deps.onNewClip(ins.id, 'image');
          deps.db.prune(deps.historySize);
        }
      } catch (e) {
        console.warn('[pipeline] image insert failed', e);
      }
    }
  };
}

File: desktop/src-tauri/src/clipboard/pipeline.rs
use super::{detect::detect_text, source_polling::ClipboardEvent, ContentType};
use crate::db::Db;
use std::sync::{Arc, Mutex};
use tokio::sync::mpsc;
use tracing::{debug, warn};

pub struct Pipeline {
    db: Arc<Mutex<Db>>,
    excluded: Vec<String>,
    history_size: i64,
    on_new_clip: Box<dyn Fn(i64, ContentType) + Send + Sync>,
}

impl Pipeline {
    pub fn new(
        db: Arc<Mutex<Db>>,
        excluded: Vec<String>,
        history_size: i64,
        on_new_clip: Box<dyn Fn(i64, ContentType) + Send + Sync>,
    ) -> Self {
        Self {
            db,
            excluded,
            history_size,
            on_new_clip,
        }
    }

    pub async fn run(
        self,
        mut rx: mpsc::Receiver<ClipboardEvent>,
        get_focused_app: impl Fn() -> Option<String>,
    ) {
        while let Some(ev) = rx.recv().await {
            let focused = get_focused_app();
            if let Some(ref app) = focused {
                if self.excluded.iter().any(|e| app.eq_ignore_ascii_case(e)) {
                    debug!("skipping capture from excluded app: {}", app);
                    continue;
                }
            }
            match ev {
                ClipboardEvent::Text { content, mime } => {
                    let ct = detect_text(&content, focused.as_deref());
                    let preview: String = content.chars().take(280).collect();
                    let mut db = self.db.lock().unwrap();
                    let now = ms_now();
                    match db.insert_clip(ct, content.as_bytes(), &mime, &preview, focused.as_deref(), now) {
                        Ok(ins) if ins.was_new => {
                            (self.on_new_clip)(ins.id, ct);
                            let _ = db.prune(self.history_size);
                        }
                        Ok(_) => debug!("dedup hit"),
                        Err(e) => warn!("insert failed: {e}"),
                    }
                }
                ClipboardEvent::Image { png_bytes } => {
                    let mut db = self.db.lock().unwrap();
                    let now = ms_now();
                    let preview = format!("Image {} bytes", png_bytes.len());
                    match db.insert_clip(
                        ContentType::Image,
                        &png_bytes,
                        "image/png",
                        &preview,
                        focused.as_deref(),
                        now,
                    ) {
                        Ok(ins) if ins.was_new => {
                            if let Ok(thumb) = crate::thumb::make_thumbnail(&png_bytes) {
                                let _ = db.set_thumbnail(ins.id, &thumb);
                            }
                            (self.on_new_clip)(ins.id, ContentType::Image);
                            let _ = db.prune(self.history_size);
                        }
                        _ => {}
                    }
                }
            }
        }
    }
}

fn ms_now() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}
